### DISCUSSION TITLE
Intrinsics instead of inline

### DISCUSSION THREAD

**Discussion by dumblob** [vlang/v#11883] [Ideas]
There has be quite some discussions over the years about having intrinsics. The recent commit https://github.com/vlang/v/commit/8650ec6916c44a119815950613c2159aecc9979f has motivated me to propose an addition of `[intrinsic]` to function definition. This would basically **guarantee** inlining unlike `[inline]` which relies on the C compiler which is (in accordance to the C standard) free to **not** inline it (as it's merely a hint, not a command).

The tag `[intrinsic]` would guarantee it by making already the *V* compiler do all the hard work of inlining the function.

Later (but still before *V 1.0*) we might consider abandoning `[inline]` altogether as it doesn't seem to provide much more than no such tag at all.

Thoughts?

**Comment by JalonSolov** [comment]
By definition, `inline` is correct, since `inline` and `intrinsic` mean different things, but that is beside the point of this discussion.  V already has `intrinsic` functions, we've just been calling them "compiler magic".

Once V starts doing optimization, it could start paying attention to the `[inline]` attribute, but I would have it treated exactly the same as the C compiler does... use it as a hint rather than an absolute directive.  Reason?  There are lots of them, and Google will show you everything you ever wanted to know.

Here's an example list from just one page I found...
```
Inline function disadvantages:
1) The added variables from the inlined function consumes additional registers, After in-lining function
if variables number which are going to use register increases than they may create overhead on register
variable resource utilization. This means that when inline function body is substituted at the point of
function call, total number of variables used by the function also gets inserted. So the number of register
going to be used for the variables will also get increased. So if after function inlining variable numbers
increase drastically then it would surely cause an overhead on register utilization.

2) If you use too many inline functions then the size of the binary executable file will be large, because of
the duplication of same code.

3) Too much inlining can also reduce your instruction cache hit rate, thus reducing the speed of instruction
fetch from that of cache memory to that of primary memory.

4) Inline function may increase compile time overhead if someone changes the code inside the inline
function then all the calling location has to be recompiled because compiler would require to replace
all the code once again to reflect the changes, otherwise it will continue with old functionality.

5) Inline functions may not be useful for many embedded systems. Because in embedded systems code
size is more important than speed.

6) Inline functions might cause thrashing because inlining might increase size of the binary executable file.
Thrashing in memory causes performance of computer to degrade.
```

**Comment by dumblob** [comment]
> By definition, `inline` is correct, since `inline` and `intrinsic` mean different things, but that is beside the point of this discussion.

Let's discuss it here :wink:. First let me clarify I'm not saying `[inline]` is incorrect. I just proposed `[intrinsic]` which could be basically seen and implemented as "guaranteed inlining". It can be implemented in other ways, but this was the one I thought about as an easy one if we want to allow programmers to define their own intrinsics.

In other words I don't care how `[intrinsic]` will be implemented, so there is no point in discussing implementation :wink:.

>V already has `intrinsic` functions, we've just been calling them "compiler magic".

Yup. This proposal is merely about allowing programmers to define intrinsics as usual functions without touching the compiler and writing compiler magic.

>Once V starts doing optimization, it could start paying attention to the [inline] attribute, but I would have it treated exactly the same as the C compiler does... use it as a hint rather than an absolute directive. Reason? There are lots of them, and Google will show you everything you ever wanted to know.

Hm, I couldn't "google out" a word about `inline` being good only as a hint and not as a directive. Could you point me at something?

The copy pasted text of "inline functions disadvantages" discusses only the (well known) worst-case implications of `inline` in apps and also makes some claims which are very specific to some compilers and do not seem to relate to (nor be valid in the context of) *V*. But it doesn't say anything about `inline` being a bad fit for being guaranteed.

**Comment by JalonSolov** [comment]
Here's one reference to inline... https://en.cppreference.com/w/c/language/inline
```
The compilers can (and usually do) ignore presence or absence of the inline specifier for the purpose of
optimization.
```

**Comment by dumblob** [comment]
That again doesn't say anything about `inline` being good **only** as a hint and bad as a "directive" (i.e. guaranteed inlining like hygienic macro expansion). I really couldn't find anything in that regard neither elsewhere.

**Comment by JalonSolov** [comment]
I never said `inline` was only good as a hint.  I said that it _is_ effectively only used as a hint, since the compiler is free to (and often does) ignore it.  It will inline whatever it's analysis says can be effectively inlined, and it won't inline others, regardless of whether you put `inline` on there or not.

**Comment by dumblob** [comment]
Ah, ok, then I misunderstood. Sure, `inline` is just a hint in C/C++ and that's also the main reason why I've created this issue where I'm proposing `[intrinsic]` :wink:.

**Comment by JalonSolov** [comment]
The problem with `intrinsic` is the meaning of `intrinsic` - it refers to standard library routines, not userland routines.

I understand what you want this for, but I disagree with adding it to V.  One of the main reasons I can think of for allowing the compiler to ignore `inline` is because most programmers simply don't use it properly.

I have seen code where every single function was declared `inline`.  I have seen code where massive _recursive_ functions were declared `inline`.

_Forcing_ the compiler to inline things when they shouldn't be would be worse than what we have now.

**Comment by dumblob** [comment]
I generally agree, but the motivation for this proposal is simply that the need arose and will presumably be reocurring due to *V* spreading among low-level programmers.

Btw. the danger of "bad" behavior holds for many other existing attributes. So it's no surprise one should do at least profiling before & after the attribute was added. And because *V* provides convenient tools to do that, it's not anything unusal to expect from a *V* programmer.
